cmd/operator/service.go

```
@@ -0,0 +1,117 @@
+package main
+
+import (
+	"context"
+	"sync"
+
+	runapi "github.com/GoogleCloudPlatform/cloud-run-release-operator/internal/run"
+	"github.com/GoogleCloudPlatform/cloud-run-release-operator/internal/util"
+	"github.com/GoogleCloudPlatform/cloud-run-release-operator/pkg/config"
+	"github.com/GoogleCloudPlatform/cloud-run-release-operator/pkg/rollout"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/api/run/v1"
+)
+
+// getTargetedServices returned a list of service records that match the target
+// configuration.
+func getTargetedServices(ctx context.Context, logger *logrus.Logger, targets []*config.Target) ([]*rollout.ServiceRecord, error) {
+	logger.Debug("filtering all targeted services")
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	var (
+		retServices []*rollout.ServiceRecord
+		retError    error
+		mu          sync.Mutex
+		wg          sync.WaitGroup
+	)
+
+	for _, target := range targets {
+		regions, err := determineRegions(ctx, logger, target)
+		if err != nil {
+			logger.Errorf("cannot determine regions: %v", err)
```

this is probably a `return errors.Wrap`.

fail early fail often.